src/core/thread/link_metrics.cpp

```diff
@@ -0,0 +1,299 @@
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for Thread link metrics.
+ */
+
+#include "link_metrics.hpp"
+
+#if OPENTHREAD_CONFIG_MLE_LINK_METRICS_ENABLE
+
+#include "common/code_utils.hpp"
+#include "common/instance.hpp"
+#include "common/locator-getters.hpp"
+
+#include "link_metrics_tlvs.hpp"
+
+namespace ot {
+
+LinkMetrics::LinkMetrics(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mLinkMetricsReportCallback(nullptr)
+    , mLinkMetricsReportCallbackContext(nullptr)
+{
+}
+
+otError LinkMetrics::LinkMetricsQuery(const otIp6Address *aDestination,
+                                      uint8_t             aSeriesId,
+                                      uint8_t *           aTypeIdFlags,
+                                      uint8_t             aTypeIdFlagsCount)
+{
+    OT_UNUSED_VARIABLE(aDestination);
+
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(aTypeIdFlagsCount <= kLinkMetricsMaxTypeIdFlags, error = OT_ERROR_INVALID_ARGS);
+
+    error = SendLinkMetricsQuery(*static_cast<const Ip6::Address *>(aDestination), aSeriesId,
+                                 reinterpret_cast<LinkMetricsTypeId *>(aTypeIdFlags), aTypeIdFlagsCount);
+
+exit:
+    return error;
+}
+
+otError LinkMetrics::AppendLinkMetricsReport(const Message &                 aMessageIn,
+                                             const Mle::LinkMetricsQueryTlv *aLinkMetricsQuery,
+                                             Message &                       aMessageOut)
```

Comment:
What about follow the convention that the first argument `aMessage` should be the target of TLV appending ? 

Also, seems only metadata of the request message is required. 

```suggestion
otError LinkMetrics::AppendLinkMetricsReport(Message &                       aMessage, 
                                             const Mle::LinkMetricsQueryTlv *aLinkMetricsQuery,
                                             const MessageMetadata &                 aRequestMetadata
                                             )
```

Reply:
Updated.